src/utils/serdes/serdes.cpp

```diff
@@ -14,121 +14,156 @@
-
-    return ret;
 }
 
-// Byte buffers serialization
-nixl_status_t nixlSerDes::addBuf(const std::string &tag, const void* buf, ssize_t len){
-
-    workingStr.append(tag);
-    workingStr.append(_bytesToString(&len, sizeof(ssize_t)));
-    workingStr.append(_bytesToString(buf, len));
-    workingStr.append("|");
+nixlDeserializer::nixlDeserializer(const std::string& str)
+    : nixlDeserializer(std::string(str))
+{}
 
-    return NIXL_SUCCESS;
+std::size_t nixlDeserializer::peekLenUnsafe(const std::size_t offset) const
+{
+    std::size_t result;
+    std::memcpy(&result, buffer_.data() + offset_ + offset, sizeof(result));
+    return result;
 }
 
-ssize_t nixlSerDes::getBufLen(const std::string &tag) const{
-    if(workingStr.compare(des_offset, tag.size(), tag) != 0){
-       //incorrect tag
-       return -1;
+std::string nixlDeserializer::getStr(const std::string_view& tag)
+{
+    if(offset_ + tag.size() + sizeof(std::size_t) > buffer_.size()) {
+        // throw std::runtime_error("deserialization data insufficient");
+        return "";
     }
+    if(buffer_.compare(offset_, tag.size(), tag) != 0) {
+        // throw std::runtime_error("deserialization tag mismatch ...");
+        return "";
+    }
+    offset_ += tag.size();
 
-    ssize_t len;
-
-    //get len
-    //_stringToBytes(&len, workingStr.data() + des_offset + tag.size(), sizeof(ssize_t));
-    _stringToBytes(&len, workingStr.substr(des_offset + tag.size(), sizeof(ssize_t)), sizeof(ssize_t));
+    const std::size_t len = peekLenUnsafe();
+    offset_ += sizeof(len);
 
-    return len;
+    if(offset_ + len + 1 > buffer_.size()) {
+        // throw std::runtime_error("deserialization data insufficient");
+        return "";
+    }
+    const std::string ret = buffer_.substr(offset_, len);
+    offset_ += len + 1;  // Also skip trailing '|'.
+    return ret;
 }
 
-nixl_status_t nixlSerDes::getBuf(const std::string &tag, void *buf, ssize_t len){
-    if(workingStr.compare(des_offset, tag.size(), tag) != 0){
-       //incorrect tag
-       return NIXL_ERR_MISMATCH;
+std::size_t nixlDeserializer::peekBufLen(const std::string_view& tag) const
+{
+    if(offset_ + tag.size() + sizeof(std::size_t) > buffer_.size()) {
+        // throw std::runtime_error("deserialization data insufficient");
+        return -1;
     }
+    if(buffer_.compare(offset_, tag.size(), tag) != 0) {
+        // throw std::runtime_error("deserialization tag mismatch ...");
+        return -1;
+    }
+    return peekLenUnsafe(tag.size());
+}
 
-    //skip over tag and size, which we assume has been read previously
-    des_offset += tag.size() + sizeof(ssize_t);
-
-    //_stringToBytes(buf, workingStr.data() + des_offset, len);
-    _stringToBytes(buf, workingStr.substr(des_offset, len), len);
-
-    //bytes in string form are twice as long, skip those plus | delimiter
-    des_offset += len + 1;
+nixl_status_t nixlDeserializer::getBuf(const std::string_view& tag, void *buf, const std::size_t len)
```

Comment:
This method seem to be used to deserialize POD integer types. I wonder whether it could be trivially "upgraded" to a template which would alleviate the need for passing len.